pkg/api/globalquery/client/stream.go

```diff
@@ -25,25 +28,61 @@ type event struct {
 
 // SSEClient is a global query client capable of streaming updates
 type SSEClient struct {
-	onUpdate StreamingUpdate
-	onFinish StreamingUpdate
+	onUpdate    StreamingUpdate
+	onFinish    StreamingUpdate
+	onKeepalive StreamingKeepalive
 
 	*client.DefaultClient
 }
 
 // StreamingUpdate is a function which operates on a received result
 type StreamingUpdate func(context.Context, *results.Result) error
 
+// StreamingKeepalive is a function which operates on a received keepalive
+type StreamingKeepalive func(context.Context) error
+
 // NewSSE creates a new streaming client for the global-query API
-func NewSSE(addr string, onUpdate, onFinish StreamingUpdate, opts ...client.Option) *SSEClient {
+func NewSSE(addr string, onUpdate, onFinish StreamingUpdate, onKeepalive StreamingKeepalive, opts ...client.Option) *SSEClient {
 	opts = append(opts, client.WithName(clientName))
 	return &SSEClient{
 		onUpdate:      onUpdate,
 		onFinish:      onFinish,
+		onKeepalive:   onKeepalive,
 		DefaultClient: client.NewDefault(addr, opts...),
 	}
 }
 
+// NewFromConfig creates the client based on cfg
+func NewSSEFromConfig(cfg *gpclient.Config, keepaliveChan chan struct{}) *SSEClient {
+	return NewSSE(cfg.Addr,
+		// TODO: this will become more informational in the future as in: printing partial results, etc.
+		func(ctx context.Context, r *results.Result) error {
+			if r == nil {
+				return nil
+			}
+			all := len(r.HostsStatuses)
+			errs := len(r.HostsStatuses.GetErrorStatuses())
+
+			logger := logging.FromContext(ctx)
+			logger.Infof("received update: %d total / %d done / %d errors", all, all-errs, errs)
+
+			return nil
+		},
+		func(ctx context.Context, r *results.Result) error { return nil },
+		func(ctx context.Context) error {
+			select {
+			case keepaliveChan <- struct{}{}:
```

Comment:
In case the channel cannot receive (e.g. because no one consumes from it) we just continue to avoid deadlocks.